### ISSUE TITLE
filter implementation improvements

### ISSUE THREAD

**Issue by aanthony1243** [influxdata/flux#1288]
Filtering in pure flux has some incorrect assumptions as laid out in #1285 .  This issue details the issues more specifically for filtering.  

1.  filter currently errors when a referenced field is missing.   This is wrong for two reasons:   
a) the filter may be designed to hit two distinct schemas as part of a disjunction, e.g.,  `r.host == "A" or r.device == "/dev/null"`.  In this case, if a table is missing `r.host`   the query should not error, and in fact can correctly match rows in tables that have a `device` column but no `host` column.  
b)  even in the case where all referenced columns are unseen, this can just be interpreted as "none of the rows in this table match".  `filter` should be considering that the next table on the stream may match.   

2.  filtering will error when types mismatch.  For example, if we filter `r._value >= 2.5` and a table in the stream has a `_value` column of type string, then the filter function evaluation fails and filtering will fail.  This could also be considered as "this row doesn't match".  

The above begs the question:  when is an error a breaking error for filter?  This is hard to define completely, but it's possible that the filter function could call another helper function that throws an error.  It's probably best to throw such an error.  Another case can be made for log-and-continue as a strategy. 

3.  Filtering should not produce empty tables.  The reason is that a filter may be designed to discard many dozens or hundreds of unwanted tables, and these should not be in the default output.  This means we will need to make special consideration for aggregates that may want to assign a default value to a known empty table.  (e.g. count(*)  or sum()|>fill(value: 0).

In summary, a missing column or a minor evaluation error should not be a cause for erroring and halting query execution.  The filter process should be able to analyze the expression for unknown references relative to a table, and determine whether the expression can still be possibly evaluated as true.

**Comment by wolffcm** [comment]
So will a predicate behave differently depending on whether it's in a filter or not?  What about this example
```
props = {
   filterOnHost: true,
}
result = tables |> filter(fn: (r) => if props.filterOnHsot then r.host == "localhost" else true)
```
The typo `filterOnHsot` is intentional.  Should this error?

**Comment by aanthony1243** [comment]
@wolffcm that actually will error for a completely different reason (no local scope for row functions!) . But even if that was fixed, we would get a similar but different error about a missing identifier.   There's a lot of cases around this that we need to understand better.  Conditionals are a good example.  I wonder if it might be solved by 3 valued logic, even if we don't expose that to the user.
